### PR TITLE
Project <tentative title>: Rework Garbage Collection.

### PR DESCRIPTION
**Pull request by FernandoS27**
This PR recreates the garbage collector using an LRU cache instead. It's not as aggressive as the older one but it gets the job done. It should be bugless.

### CODE REVIEW

**Comment by goldenx86** [comment]
Project Thanatos?

**Comment by FernandoS27** [comment]
> Project Thanatos?

the current name is final!

**Comment by goldenx86** [comment]
"Project <tentative title> & Knuckles" it is then.

**Comment by v1993** [comment]
Next progress report is gonna be awesome

**Comment by Lt-knb** [comment]
While this pr fixes #6849, the crashes of #5307 now also happen on Vulkan because this new Garbage Collector is apparently not able to get rid of them in comparison to the previous one. This is the only regression so far

File: src/common/lru_cache.h [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+            last_item = item.prev;
+            if (last_item) {
+                last_item->next = nullptr;
+            }
+        }
+    }
+
+    std::deque<Item> item_pool;
+    std::deque<size_t> free_items;
+    Item* first_item;
```
Comment (lioncash) [inline src/common/lru_cache.h:137]:
```suggestion
    Item* first_item{};
```

Ditto below. Then the constructor can be defaulted.

File: src/common/lru_cache.h [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+            }
+            iterator = next;
+        }
+    }
+
+private:
+    size_t build() {
+        if (free_items.empty()) {
+            const size_t item_id = item_pool.size();
+            item_pool.emplace_back();
```
Comment (lioncash) [inline src/common/lru_cache.h:86]:
```suggestion
            auto& item = item_pool.emplace_back();
```

File: src/video_core/buffer_cache/buffer_base.h [hunk 3]
```diff
@@ -307,6 +297,8 @@ class BufferBase {
         return words.size_bytes;
     }
 
+    size_t lru_id;
```
Comment (lioncash) [inline src/video_core/buffer_cache/buffer_base.h:300]: This should be private and have an API around it to be consistent with the rest of the class.

File: src/video_core/texture_cache/image_base.h [hunk 4]
```diff
@@ -80,7 +80,7 @@ struct ImageBase {
     VAddr cpu_addr_end = 0;
 
     u64 modification_tick = 0;
-    u64 frame_tick = 0;
+    size_t lru_index = ~0;
```
Comment (lioncash) [inline src/video_core/texture_cache/image_base.h:83]:
```suggestion
    size_t lru_index = SIZE_MAX;
```

File: src/common/lru_cache.h [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+                }
+            } else {
+                func(iterator->obj);
+            }
+            iterator = next;
+        }
+    }
+
+private:
+    size_t build() {
```
Comment (lioncash) [inline src/common/lru_cache.h:83]:
```suggestion
    size_t Build() {
```

File: src/common/lru_cache.h [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+        }
+        const size_t item_id = free_items.front();
+        free_items.pop_front();
+        auto& item = item_pool[item_id];
+        item.next = nullptr;
+        item.prev = nullptr;
+        return item_id;
+    }
+
+    void attach(Item& item) {
```
Comment (lioncash) [inline src/common/lru_cache.h:100]:
```suggestion
    void Attach(Item& item) {
```

File: src/common/lru_cache.h [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+            last_item = &item;
+        } else {
+            item.prev = last_item;
+            last_item->next = &item;
+            item.next = nullptr;
+            last_item = &item;
+        }
+    }
+
+    void detach(Item& item) {
```
Comment (lioncash) [inline src/common/lru_cache.h:114]:
```suggestion
    void Detach(Item& item) {
```

**Comment by Lt-knb** [comment]
I don't know what the new commit does but i tested EA2023 and the behavior remains the same in Splatoon 2

**Review by bunnei — APPROVED** [review]

**Pull request merged**